Update installation instructions (#1637)
cocotb/cocotb: documentation/source/quickstart.rst

Installing cocotb
=================

> Pre-requisites Installation

Reviewer: ```suggestion
Installation of Pre-requisites
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^

`Conda <https://conda.io/>`_ is an open-source package management system and environment management system that runs on Windows, macOS and Linux.

Download and install `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_ from https://conda.io/ according to your OS and required architecture.

You will also need to install a compiler (GCC or Clang) and GNU Make.

On Windows you can do this easily do this with conda:

.. code-block:: bash

    conda install -c msys2 m2-base m2-make m2w64-toolchain libpython

Native Debian/Ubuntu-based
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: bash

    sudo apt-get install make gcc g++ swig python3 python3-dev python3-pip

Native Red Hat-based
^^^^^^^^^^^^^^^^^^^^

.. code-block:: bash

    sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip

macOS Packages
^^^^^^^^^^^^^^

You need a few packages installed to get cocotb running on macOS.
Installing a package manager really helps things out here.

`Brew <https://brew.sh/>`_ seems to be the most popular, so we'll assume you have that installed.

.. code-block:: bash

    brew install python icarus-verilog gtkwave

> For more installation options (also 32-bit simulator) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.

Reviewer: ```suggestion
For more installation options (also for 32-bit simulators) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.
```

Simulator Installation
^^^^^^^^^^^^^^^^^^^^^^

For details and the list of supported simulators see :ref:`_simulator-support`.

> Install and ensure that simulator executable is in the ``PATH``.   

Reviewer: ```suggestion
Install and ensure that the simulator executable is on the ``PATH``.   
```

> .. _installation_via_pip:

Reviewer: ```suggestion
.. _installation_via_pip::
```

.. versionchanged:: 1.4 Dropped Python 2 support

> .. note:: In order to use a 32bit simulator you need to use a 32bit version of Python.

Reviewer 2: ```suggestion
.. note:: In order to use a 32-bit simulator you need to use a 32-bit version of Python.
```

Installation with conda (all OS)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`Conda <https://conda.io/>`_ is an open-source package and environment management system that runs on Windows, macOS and Linux.

Download and install `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_ from https://conda.io/ for your OS and architecture.

You will also need to install a compiler (GCC or Clang) and GNU Make.

On Windows you can easily do this with conda:

.. code-block:: bash

    conda install -c msys2 m2-base m2-make m2w64-toolchain libpython

Native Installation for Debian/Ubuntu-based Systems
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: bash

>     sudo apt-get install make gcc g++ swig python3 python3-dev python3-pip

Reviewer 2: ```suggestion
    sudo apt-get install make gcc g++ python3 python3-dev python3-pip
    sudo apt-get install swig  # optional, needed for one of the examples
```

Native Installation for Red Hat-based Systems
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: bash

>     sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip

Reviewer 2: ```suggestion
    sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip
    sudo yum install swig  # optional, needed for one of the examples
```

> macOS Packages
> ^^^^^^^^^^^^^^

Reviewer 2: For consistency with the other headings:
```suggestion
Installation for macOS
^^^^^^^^^^^^^^^^^^^^^^
```

You need a few packages installed to get cocotb running on macOS.
Installing a package manager really helps things out here.

`Brew <https://brew.sh/>`_ seems to be the most popular, so we'll assume you have that installed.

.. code-block:: bash

    brew install python icarus-verilog gtkwave

.. seealso::

>    For more installation options (also for 32-bit simulators) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.

Reviewer 2: I think this should be below the `Simulator Installation` heading

Author: I have changed this a bit to remove `simulators`  since this was the intention. Feel free to change.

    brew install python icarus-verilog gtkwave

Red Hat-based
^^^^^^^^^^^^^
.. seealso::

.. code-block:: bash
   For more installation options (also for 32-bit simulators) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.

    sudo yum install gcc gcc-c++ libstdc++-devel swig python-devel

Simulator Installation
^^^^^^^^^^^^^^^^^^^^^^

Windows Installation
--------------------
For detailed instructions and the list of supported simulators see :ref:`simulator-support`.

Download the MinGW installer from https://osdn.net/projects/mingw/releases/.

Run the GUI installer and specify a directory you would like the environment
installed in. The installer will retrieve a list of possible packages, when this
is done press "Continue". The MinGW Installation Manager is then launched.
.. note::

The following packages need selecting by checking the tick box and selecting
"Mark for installation"
    Ensure that path to the simulator executable is appended to the ``PATH`` environment variable.

.. code-block:: bash

    Basic Installation
      -- mingw-developer-tools
      -- mingw32-base
      -- mingw32-gcc-g++

>       -- msys-base

Reviewer 2: Are we sure we want to remove these instructions? Maybe worth putting them on the wiki first, so they don't get lost completely.

Author: I think yes. On the wiki, I can add msys2 instructions.

Author: Added instruction for MSYS/MINGW here: https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions#windwows-installation-using-msys2

>     pip install cocotb  # use `pip3` in case multiple of Python/pip version installed

Reviewer 2: ```suggestion
    pip install cocotb  # use `pip3` in case multiple versions of Python or pip are installed
```

Author: Maybe even better would be ``use `pip3` if available`` ?

Author: Or better opposite 
```bash
pip3 install cocotb  # use `pip` if `pip3` not available
```

Reviewer 2: How about we just use
```
pip install cocotb
```
but add a note before the first time we mention `pip` like
```
.. warning::
    ``pip`` may belong to a different python installation to what you expect.
    Use ``pip -V`` to check.
    If this shows Python 2.7, use ``pip3`` or ``python3 -m pip`` in place of ``pip`` in the commands below.
```

Author: Sound OK. Yo sure **before** or rather close to `cocotb-config` note?

Reviewer 2: You're right, your suggested placement makes more sense.

Author: :+1: If we see issues we move up ;-)

>     For user local installation follow the `pip User Guide <https://pip.pypa.io/en/stable/user_guide/#user-installs/>`_.

Reviewer 2: ```suggestion
    For user-local installation, follow the `pip User Guide <https://pip.pypa.io/en/stable/user_guide/#user-installs/>`_.
```

>     pip install https://github.com/cocotb/cocotb/archive/master.zip  # use `pip3` in case multiple of Python/pip version installed

Reviewer 2: ```suggestion
    pip install https://github.com/cocotb/cocotb/archive/master.zip  # use `pip3` in case multiple versions of Python or pip are installed
```
